sdk/common/src/test/java/io/opentelemetry/sdk/internal/DaemonThreadFactoryTest.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.internal;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.clearInvocations;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoInteractions;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import org.junit.jupiter.api.Test;
+
+class DaemonThreadFactoryTest {
+
+  @Test
+  void verifyUncaughtExceptions()
+      throws ExecutionException, InterruptedException, TimeoutException {
+    Thread.UncaughtExceptionHandler defaultHandler = mock();
+    ThreadFactory delegateFactory =
+        r -> {
+          Thread thread = Executors.defaultThreadFactory().newThread(r);
+          thread.setUncaughtExceptionHandler(defaultHandler);
+          return thread;
+        };
+    ExecutorService service =
+        Executors.newSingleThreadExecutor(new DaemonThreadFactory("test", delegateFactory));
+    Runnable runnable =
+        () -> {
+          Thread.UncaughtExceptionHandler uncaughtExceptionHandler =
+              Thread.currentThread().getUncaughtExceptionHandler();
+
+          assertThat(uncaughtExceptionHandler)
+              .isInstanceOf(DaemonThreadFactory.ManagedUncaughtExceptionHandler.class);
+
+          Thread threadMock = mock();
+
+          // Verify interrupted exception
+          uncaughtExceptionHandler.uncaughtException(threadMock, new InterruptedException());
+          verify(threadMock).interrupt();
+          verifyNoInteractions(defaultHandler);
+
+          // Verify delegate exception
+          clearInvocations(threadMock, defaultHandler);
+          IllegalStateException e = new IllegalStateException();
+          uncaughtExceptionHandler.uncaughtException(threadMock, e);
+          verify(defaultHandler).uncaughtException(threadMock, e);
```

Comment:
I think that doing all assertions/verifications in an async block when the main test body doesn't also verify the execution is a code smell.

I think there could be something at the end of the `Runnable` (AtomicBoolean, CountdownLatch, etc) that you can then verify in the main thread. That way the reader can easily determine that the runnable completed which means the assertions all ran and passed.